src/Servers.jl

```diff
@@ -69,15 +69,20 @@ function check_rate_limit(tcp, rate_limit::Rational{Int})
 end
 
 "Convenience object for passing around server details"
-struct Server{S <: Union{SSLConfig, Nothing}, I <: Base.IOServer}
+struct Server{S <: Union{SSLConfig, Nothing}, I <: Base.IOServer, F <: Union{Function, Vector{Function}, Nothing}}
     ssl::S
     server::I
     hostname::String
     hostport::String
+    on_shutdown::F
 end
 
 Base.isopen(s::Server) = isopen(s.server)
-Base.close(s::Server) = close(s.server)
+Base.close(s::Server) = (shutdown(s.on_shutdown); close(s.server))
```

Comment:
Would it be useful to pass `s` to the shutdown functions?

Reply:
If there's a possibility we might want to do other (not specified by the user) shutdown tasks that needed the other fields this would make sense? If not I think it's cleaner as it is for now, and it would always be easy to add it in the future with another method for `shutdown`.

Reply:
It was just a thought that maybe you need the server object for shutting down your own stuff too, but yea can always be added later by checking `applicable` I think.

Reply:
Oh yeah `applicable` would let you do that, good idea. It's probably worth having the extra functionality in? I'll add it along with some tests and documentation.

Reply:
I don't have a usecase for it from top of my head, so maybe it can wait. Let's see what others have to say :)

Reply:
In the past, I have built similar functionality in a wrapper around `Server` in a custom framework. I'm not sure if this use case would still be applicable here, but I used the shutdown functionality there to correctly close any open WebSockets instead of just destroying them (which can leave browsers confused).

Reply:
Yeah that sounds like a good use case for this - are WebSockets destroyed when `close(s.server)` is performed?

Reply:
Yes, just prior to it. The `stop` function for the wrapper, which keeps track of open WebSockets, simply iterates over all the WebSockets that are still open, initiates closing them, waits for them to actually close and then calls `close(wrapper.server)`.

Quite happy to see these types of changes making it into HTTP.jl. The main reason for this wrapper to exist, is for purposes such as this (and #602 which has also been fixed). Thanks for putting in the work!